File: backend/app/models/operations.py
from __future__ import annotations

from app.models.base import (
    GUID,
    JSON,
    Any,
    Base,
    Boolean,
    CheckConstraint,
    Float,
    ForeignKey,
    Integer,
    Mapped,
    String,
    Text,
    TimestampMixin,
    UniqueConstraint,
    UTCDateTime,
    dt,
    mapped_column,
    utcnow,
    uuid,
)


class OperationalMetricSnapshot(Base):
    __tablename__ = "operational_metric_snapshots"

    id: Mapped[uuid.UUID] = mapped_column(GUID(), primary_key=True, default=uuid.uuid4)
    bucket_started_at: Mapped[dt.datetime] = mapped_column(
        UTCDateTime(), nullable=False, unique=True, index=True
    )
    generated_at: Mapped[dt.datetime] = mapped_column(UTCDateTime(), nullable=False, index=True)
    expires_at: Mapped[dt.datetime] = mapped_column(UTCDateTime(), nullable=False, index=True)
    health: Mapped[str] = mapped_column(String(20), nullable=False, index=True)
    schema_version: Mapped[str] = mapped_column(String(100), nullable=False)
    sample_count: Mapped[int] = mapped_column(Integer, nullable=False)
    alert_count: Mapped[int] = mapped_column(Integer, nullable=False)
    content_hash: Mapped[str] = mapped_column(String(64), nullable=False, unique=True, index=True)
    collected_by: Mapped[str] = mapped_column(String(160), nullable=False)
    created_at: Mapped[dt.datetime] = mapped_column(UTCDateTime(), nullable=False, default=utcnow)

    __table_args__ = (
        CheckConstraint(
            "health IN ('healthy', 'degraded', 'critical')",
            name="ck_operational_metric_snapshots_health",
        ),
        CheckConstraint(
            "sample_count >= 0 AND alert_count >= 0",
            name="ck_operational_metric_snapshots_counts",
        ),
        CheckConstraint(
            "expires_at > generated_at",
            name="ck_operational_metric_snapshots_expiry",
        ),
    )

class OperationalMetricPoint(Base):
    __tablename__ = "operational_metric_points"

    id: Mapped[uuid.UUID] = mapped_column(GUID(), primary_key=True, default=uuid.uuid4)
    snapshot_id: Mapped[uuid.UUID] = mapped_column(
        GUID(),
        ForeignKey("operational_metric_snapshots.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    metric_name: Mapped[str] = mapped_column(String(180), nullable=False, index=True)
    metric_value: Mapped[float] = mapped_column(Float, nullable=False)
    labels_json: Mapped[dict[str, Any]] = mapped_column(JSON, nullable=False, default=dict)
    labels_hash: Mapped[str] = mapped_column(String(64), nullable=False)
    help_text: Mapped[str] = mapped_column(Text, nullable=False)
    recorded_at: Mapped[dt.datetime] = mapped_column(UTCDateTime(), nullable=False, index=True)
    created_at: Mapped[dt.datetime] = mapped_column(UTCDateTime(), nullable=False, default=utcnow)

    __table_args__ = (
        UniqueConstraint(
            "snapshot_id",
            "metric_name",
            "labels_hash",
            name="uq_operational_metric_points_snapshot_metric_labels",
        ),
    )

class OperationalSLOEvaluation(Base):
    __tablename__ = "operational_slo_evaluations"

    id: Mapped[uuid.UUID] = mapped_column(GUID(), primary_key=True, default=uuid.uuid4)
    snapshot_id: Mapped[uuid.UUID] = mapped_column(
        GUID(),
        ForeignKey("operational_metric_snapshots.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    slo_key: Mapped[str] = mapped_column(String(120), nullable=False, index=True)
    scope: Mapped[str] = mapped_column(String(40), nullable=False, index=True)
    status: Mapped[str] = mapped_column(String(30), nullable=False, index=True)
    target_ratio: Mapped[float] = mapped_column(Float, nullable=False)
    observed_ratio: Mapped[float | None] = mapped_column(Float)
    error_budget_remaining_ratio: Mapped[float | None] = mapped_column(Float)
    window_seconds: Mapped[int] = mapped_column(Integer, nullable=False)
    short_window_seconds: Mapped[int] = mapped_column(Integer, nullable=False)
    short_observed_ratio: Mapped[float | None] = mapped_column(Float)
    burn_rate: Mapped[float | None] = mapped_column(Float)
    short_burn_rate: Mapped[float | None] = mapped_column(Float)
    burn_alert_level: Mapped[str] = mapped_column(
        String(20), nullable=False, default="none", index=True
    )
    sample_count: Mapped[int] = mapped_column(Integer, nullable=False)
    good_sample_count: Mapped[int] = mapped_column(Integer, nullable=False)
    window_started_at: Mapped[dt.datetime] = mapped_column(UTCDateTime(), nullable=False)
    window_ended_at: Mapped[dt.datetime] = mapped_column(UTCDateTime(), nullable=False)
    evaluated_at: Mapped[dt.datetime] = mapped_column(UTCDateTime(), nullable=False, index=True)
    details_json: Mapped[dict[str, Any]] = mapped_column(JSON, nullable=False, default=dict)
    evaluation_hash: Mapped[str] = mapped_column(
        String(64), nullable=False, unique=True, index=True
    )
    created_at: Mapped[dt.datetime] = mapped_column(UTCDateTime(), nullable=False, default=utcnow)

    __table_args__ = (
        UniqueConstraint(
            "snapshot_id",
            "slo_key",
            name="uq_operational_slo_evaluations_snapshot_key",
        ),
        CheckConstraint(
            "scope IN ('identity', 'worker')",
            name="ck_operational_slo_evaluations_scope",
        ),
        CheckConstraint(
            "status IN ('met', 'breached', 'insufficient')",
            name="ck_operational_slo_evaluations_status",
        ),
        CheckConstraint(
            "burn_alert_level IN ('none', 'warning', 'critical')",
            name="ck_operational_slo_evaluations_burn_alert_level",
        ),
        CheckConstraint(
            "target_ratio >= 0 AND target_ratio <= 1 "
            "AND (observed_ratio IS NULL OR "
            "(observed_ratio >= 0 AND observed_ratio <= 1)) "
            "AND (short_observed_ratio IS NULL OR "
            "(short_observed_ratio >= 0 AND short_observed_ratio <= 1)) "
            "AND (error_budget_remaining_ratio IS NULL OR "
            "(error_budget_remaining_ratio >= 0 "
            "AND error_budget_remaining_ratio <= 1)) "
            "AND (burn_rate IS NULL OR burn_rate >= 0) "
            "AND (short_burn_rate IS NULL OR short_burn_rate >= 0)",
            name="ck_operational_slo_evaluations_ratios",
        ),
        CheckConstraint(
            "window_seconds > 0 AND short_window_seconds > 0 "
            "AND short_window_seconds <= window_seconds AND sample_count >= 0 "
            "AND good_sample_count >= 0 AND good_sample_count <= sample_count",
            name="ck_operational_slo_evaluations_counts",
        ),
    )

class OperationalAlertIncident(TimestampMixin, Base):
    __tablename__ = "operational_alert_incidents"

    id: Mapped[uuid.UUID] = mapped_column(GUID(), primary_key=True, default=uuid.uuid4)
    active_key: Mapped[str | None] = mapped_column(String(180), unique=True, index=True)
    alert_key: Mapped[str] = mapped_column(String(180), nullable=False, index=True)
    source_type: Mapped[str] = mapped_column(String(30), nullable=False, index=True)
    severity: Mapped[str] = mapped_column(String(20), nullable=False, index=True)
    route: Mapped[str] = mapped_column(String(120), nullable=False)
    summary: Mapped[str] = mapped_column(String(500), nullable=False)
    metric_name: Mapped[str | None] = mapped_column(String(180), index=True)
    current_value: Mapped[float | None] = mapped_column(Float)
    threshold: Mapped[float | None] = mapped_column(Float)
    slo_evaluation_id: Mapped[uuid.UUID | None] = mapped_column(
        GUID(),
        ForeignKey("operational_slo_evaluations.id", ondelete="SET NULL"),
        index=True,
    )
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="open", index=True)
    opened_at: Mapped[dt.datetime] = mapped_column(UTCDateTime(), nullable=False, index=True)
    last_seen_at: Mapped[dt.datetime] = mapped_column(UTCDateTime(), nullable=False, index=True)
    resolved_at: Mapped[dt.datetime | None] = mapped_column(UTCDateTime(), index=True)
    acknowledged_at: Mapped[dt.datetime | None] = mapped_column(UTCDateTime(), index=True)
    acknowledged_by_identity_json: Mapped[dict[str, Any] | None] = mapped_column(JSON)
    assigned_to: Mapped[str | None] = mapped_column(String(160), index=True)
    escalation_level: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    occurrence_count: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    transition_version: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    latest_context_json: Mapped[dict[str, Any]] = mapped_column(JSON, nullable=False, default=dict)

    __table_args__ = (
        CheckConstraint(
            "source_type IN ('derived_alert', 'slo', 'test')",
            name="ck_operational_alert_incidents_source",
        ),
        CheckConstraint(
            "severity IN ('warning', 'critical')",
            name="ck_operational_alert_incidents_severity",
        ),
        CheckConstraint(
            "status IN ('open', 'resolved')",
            name="ck_operational_alert_incidents_status",
        ),
        CheckConstraint(
            "occurrence_count > 0 AND transition_version > 0 "
            "AND escalation_level >= 0 AND escalation_level <= 3",
            name="ck_operational_alert_incidents_counts",
        ),
        CheckConstraint(
            "(status = 'open' AND active_key IS NOT NULL "
            "AND resolved_at IS NULL) OR "
            "(status = 'resolved' AND active_key IS NULL "
            "AND resolved_at IS NOT NULL)",
            name="ck_operational_alert_incidents_lifecycle",
        ),
    )

class OperationalAlertIncidentAction(Base):
    __tablename__ = "operational_alert_incident_actions"

    id: Mapped[uuid.UUID] = mapped_column(GUID(), primary_key=True, default=uuid.uuid4)
    incident_id: Mapped[uuid.UUID] = mapped_column(
        GUID(),
        ForeignKey("operational_alert_incidents.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    action_type: Mapped[str] = mapped_column(String(30), nullable=False, index=True)
    reason: Mapped[str] = mapped_column(String(500), nullable=False)
    assignee: Mapped[str | None] = mapped_column(String(160))
    actor_identity_json: Mapped[dict[str, Any]] = mapped_column(JSON, nullable=False)
    identity_verified: Mapped[bool] = mapped_column(Boolean, nullable=False)
    occurred_at: Mapped[dt.datetime] = mapped_column(UTCDateTime(), nullable=False, index=True)
    previous_action_hash: Mapped[str | None] = mapped_column(String(64))
    action_hash: Mapped[str] = mapped_column(String(64), nullable=False, unique=True, index=True)
    created_at: Mapped[dt.datetime] = mapped_column(UTCDateTime(), nullable=False, default=utcnow)

    __table_args__ = (
        CheckConstraint(
            "action_type IN ('acknowledged', 'assigned', 'note', 'escalated')",
            name="ck_operational_alert_incident_actions_type",
        ),
    )

class OperationalAlertDelivery(TimestampMixin, Base):
    __tablename__ = "operational_alert_deliveries"

    id: Mapped[uuid.UUID] = mapped_column(GUID(), primary_key=True, default=uuid.uuid4)
    incident_id: Mapped[uuid.UUID] = mapped_column(
        GUID(),
        ForeignKey("operational_alert_incidents.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    delivery_job_id: Mapped[uuid.UUID | None] = mapped_column(
        GUID(),
        ForeignKey("agent_execution_jobs.id", ondelete="SET NULL"),
        unique=True,
        index=True,
    )
    transition_type: Mapped[str] = mapped_column(String(20), nullable=False, index=True)
    transition_version: Mapped[int] = mapped_column(Integer, nullable=False)
    destination_hash: Mapped[str] = mapped_column(String(64), nullable=False)
    signing_key_id: Mapped[str] = mapped_column(String(120), nullable=False)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="queued", index=True)
    payload_json: Mapped[dict[str, Any]] = mapped_column(JSON, nullable=False)
    payload_hash: Mapped[str] = mapped_column(String(64), nullable=False, unique=True, index=True)
    requested_at: Mapped[dt.datetime] = mapped_column(UTCDateTime(), nullable=False, index=True)
    last_attempt_at: Mapped[dt.datetime | None] = mapped_column(UTCDateTime(), index=True)
    attempt_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    response_status: Mapped[int | None] = mapped_column(Integer)
    response_hash: Mapped[str | None] = mapped_column(String(64))
    provider_name: Mapped[str | None] = mapped_column(String(80), index=True)
    provider_event_id: Mapped[str | None] = mapped_column(String(200), index=True)
    provider_receipt_id: Mapped[str | None] = mapped_column(String(200), index=True)
    provider_accepted_at: Mapped[dt.datetime | None] = mapped_column(UTCDateTime(), index=True)
    error_message: Mapped[str | None] = mapped_column(Text)
    delivered_at: Mapped[dt.datetime | None] = mapped_column(UTCDateTime(), index=True)

    __table_args__ = (
        UniqueConstraint(
            "incident_id",
            "transition_type",
            "transition_version",
            name="uq_operational_alert_deliveries_transition",
        ),
        CheckConstraint(
            "transition_type IN ('opened', 'resolved', 'test', 'escalated')",
            name="ck_operational_alert_deliveries_transition",
        ),
        CheckConstraint(
            "status IN ('queued', 'delivered', 'failed')",
            name="ck_operational_alert_deliveries_status",
        ),
        CheckConstraint(
            "transition_version > 0 AND attempt_count >= 0",
            name="ck_operational_alert_deliveries_attempts",
        ),
    )

File: frontend/types/api/operations.ts
import type { UUID } from "./common";



export type OperationalMetricSample = {
  name: string;
  value: number;
  metric_type: "gauge";
  help: string;
  labels: Record<string, string>;
};

export type OperationalAlert = {
  key: string;
  severity: "warning" | "critical";
  metric_name: string;
  current_value: number;
  threshold: number;
  summary: string;
  route: string;
};

export type OperationalMetrics = {
  schema_version: string;
  generated_at: string;
  health: "healthy" | "degraded" | "critical";
  samples: OperationalMetricSample[];
  alerts: OperationalAlert[];
};

export type OperationalReliabilityPermission = {
  policy_version: string;
  can_audit: boolean;
  can_administer: boolean;
  auditor_roles: string[];
  administrator_roles: string[];
  identity_verified: boolean;
  signer_role: string | null;
};

export type OperationalReliabilityPolicy = {
  snapshot_enabled: boolean;
  snapshot_interval_seconds: number;
  snapshot_retention_days: number;
  slo_window_seconds: number;
  slo_short_window_seconds: number;
  slo_min_samples: number;
  identity_slo_target: number;
  worker_slo_target: number;
  slo_warning_burn_rate: number;
  slo_critical_burn_rate: number;
  incident_escalation_seconds: number;
  paging_enabled: boolean;
  paging_destination: string | null;
  paging_transport: string;
  paging_active_key_id: string | null;
  paging_tls_verified: boolean;
  paging_provider: string;
  paging_receipt_required: boolean;
  paging_secret_source:
    | "projected_file"
    | "environment_keyring"
    | "legacy_environment"
    | "unconfigured"
    | "invalid";
  paging_key_count: number;
};

export type OperationalReadinessCheck = {
  key: string;
  status: "passed" | "failed";
  summary: string;
  evidence: string | null;
};

export type OperationalStagingReadiness = {
  schema_version: string;
  ready: boolean;
  passed_count: number;
  failed_count: number;
  latest_receipt_delivery_id: UUID | null;
  checks: OperationalReadinessCheck[];
};

export type OperationalSnapshotSummary = {
  id: UUID;
  bucket_started_at: string;
  generated_at: string;
  expires_at: string;
  health: "healthy" | "degraded" | "critical";
  sample_count: number;
  alert_count: number;
  content_hash: string;
  worker_online: number;
  queue_depth: number;
  identity_retention_due: number;
  identity_inactive_provider_token: number;
};

export type OperationalSLOEvaluation = {
  id: UUID;
  snapshot_id: UUID;
  slo_key: string;
  scope: "identity" | "worker";
  status: "met" | "breached" | "insufficient";
  target_ratio: number;
  observed_ratio: number | null;
  error_budget_remaining_ratio: number | null;
  window_seconds: number;
  short_window_seconds: number;
  short_observed_ratio: number | null;
  burn_rate: number | null;
  short_burn_rate: number | null;
  burn_alert_level: "none" | "warning" | "critical";
  sample_count: number;
  good_sample_count: number;
  window_started_at: string;
  window_ended_at: string;
  evaluated_at: string;
  details_json: Record<string, unknown>;
  evaluation_hash: string;
};

export type OperationalAlertIncident = {
  id: UUID;
  alert_key: string;
  source_type: "derived_alert" | "slo" | "test";
  severity: "warning" | "critical";
  route: string;
  summary: string;
  metric_name: string | null;
  current_value: number | null;
  threshold: number | null;
  slo_evaluation_id: UUID | null;
  status: "open" | "resolved";
  opened_at: string;
  last_seen_at: string;
  resolved_at: string | null;
  acknowledged_at: string | null;
  acknowledged_by_identity_json: Record<string, unknown> | null;
  assigned_to: string | null;
  escalation_level: number;
  occurrence_count: number;
  transition_version: number;
};

export type OperationalAlertIncidentAction = {
  id: UUID;
  incident_id: UUID;
  action_type: "acknowledged" | "assigned" | "note" | "escalated";
  reason: string;
  assignee: string | null;
  actor_identity_json: Record<string, unknown>;
  identity_verified: boolean;
  occurred_at: string;
  previous_action_hash: string | null;
  action_hash: string;
};

export type OperationalAlertDelivery = {
  id: UUID;
  incident_id: UUID;
  delivery_job_id: UUID | null;
  transition_type: "opened" | "resolved" | "test" | "escalated";
  transition_version: number;
  destination_fingerprint: string;
  signing_key_id: string;
  status: "queued" | "delivered" | "failed";
  payload_hash: string;
  requested_at: string;
  last_attempt_at: string | null;
  attempt_count: number;
  response_status: number | null;
  response_hash: string | null;
  provider_name: string | null;
  provider_event_id: string | null;
  provider_receipt_id: string | null;
  provider_accepted_at: string | null;
  error_message: string | null;
  delivered_at: string | null;
};

export type OperationalReliabilityOverview = {
  schema_version: string;
  generated_at: string;
  health: "healthy" | "degraded" | "critical";
  snapshot_count: number;
  open_incident_count: number;
  delivery_status_counts: Record<string, number>;
  latest_snapshot: OperationalSnapshotSummary | null;
  latest_slo_evaluations: OperationalSLOEvaluation[];
  snapshots: OperationalSnapshotSummary[];
  incidents: OperationalAlertIncident[];
  incident_actions: OperationalAlertIncidentAction[];
  deliveries: OperationalAlertDelivery[];
  policy: OperationalReliabilityPolicy;
  staging_readiness: OperationalStagingReadiness;
  permissions: OperationalReliabilityPermission;
};

export type IsolationPolicy = {
  policy_id: string;
  policy_version: string;
  display_name: string;
  workload_kinds: Array<"tool" | "rag">;
  network_mode: "none" | "internal_allowlist";
  allowed_network_hosts: string[];
  credential_mode: "none" | "environment_allowlist";
  allowed_secret_envs: string[];
  filesystem_mode: "ephemeral" | "read_only";
  read_only_paths: string[];
  writable_paths: string[];
  allowed_tools: string[];
  subprocess_allowed: boolean;
  max_execution_seconds: number;
  max_output_bytes: number;
};

export type IsolationPolicyRegistry = {
  registry_version: string;
  policy_count: number;
  policies: IsolationPolicy[];
};
